test/functional/url-expander/test-expander.js

```diff
@@ -0,0 +1,870 @@
+/**
+ * Copyright 2018 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {Expander} from '../../../src/service/url-expander/expander';
+import {GlobalVariableSource} from '../../../src/service/url-replacements-impl';
+import {Services} from '../../../src/services';
+import {WindowInterface} from '../../../src/window-interface';
+import {parseUrl} from '../../../src/url';
+import {setCookie} from '../../../src/cookies';
+import * as trackPromise from '../../../src/impression';
+
+describes.realWin('Expander', {
+  amp: {
+    ampdoc: 'single',
+  },
+}, env => {
+
+  let expander;
+  let variableSource;
+  let win;
+
+  beforeEach(() => {
+    win = env.win;
+    variableSource = new GlobalVariableSource(env.ampdoc);
+    expander = new Expander(variableSource);
+  });
+
+  afterEach(() => {
+    sandbox.restore();
+  });
+
+  describe('#eliminateOverlaps', () => {
+    const url = 'http://www.google.com/?client=CLIENT_ID(__ga)&canon=CANONICAL_URL&random=RANDOM';
+
+    it('should handle empty', () => {
+      const array = [];
+      const expected = null;
+      expect(expander.eliminateOverlaps_(array, url)).to.deep.equal(expected);
```

Comment:
we prefer to test public method only, so if later on we changed the internal implementation but not the public API, we're confident if it passes the current tests.

in this case, you need to come up with some test case that have overlaps in the URL, and test the end result of `expand()`

Reply:
we discussed this one in person, but tests were changed. Also pretty much made irrelevant with regex behavior